Add more node utility extensions for update rules

reaperrr opened the pull request:
Adds `MoveNode`, `MoveAndRenameNode` and `RenameChildrenMatching` extensions.

Only updated 4 rules to serve as examples, but #15010 has some prime candidates that could benefit a lot from this (in particular, `ReworkCheckboxes` and `SplitGateFromBuilding`).

For the record, I tested all new extensions and updated rules.

reaperrr committed:
Refactor some update util extensions

Using arguments instead of separate overloads, plus better support for
automatically handling trait/property removals ('-' prefix).

reaperrr committed:
Use new update util extensions to simplify some update rules

reaperrr committed:
Replace spaces with tabs in RenameEmitInfantryOnSell

reaperrr commented:
Note: Some rules in #15010 could be shortened a lot and as result become a bit easier to review if this PR could get reviewed and merged quickly.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/ChangeIntensityToDuration.cs:
```diff
@@ -30,13 +30,7 @@ public override string Description
-
-				intensity.RenameKeyPreservingSuffix("Duration");
-			}
+				sod.RenameLastChildMatching("Intensity", "Duration");
```
When renaming nodes we really should be doing everything, not just the last one.  Otherwise the update will leave  you in a broken state if you're renaming a trait node and an actor has multiple.

reaperrr commented on OpenRA.Mods.Common/UpdateRules/UpdateUtils.cs:
```diff
@@ -272,6 +272,18 @@ public static void RemoveNode(this MiniYamlNode node, MiniYamlNode toRemove)
+			fromNode.Value.Nodes.Remove(node);
+		}
+
+		public static void MoveAndRenameNode(this MiniYamlNode node, MiniYamlNode fromNode, MiniYamlNode toNode, string newKey)
```
Made preserving suffix optional instead (defaulting to "true", though).

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/RenameEmitInfantryOnSell.cs:
```diff
@@ -13,33 +13,32 @@
-                uneios.RenameKeyPreservingSuffix("-SpawnActorsOnSell");
+		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		{
+			actorNode.RenameChildrenMatching("-EmitInfantryOnSell", "-SpawnActorsOnSell");
```
We are generally really bad about (not) renaming removals in most cases.  How about we add a `bool includeRemovals=true` parameter to the `*Rename*` helpers so that we don't need to worry about it?

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/RenameEmitInfantryOnSell.cs:
```diff
@@ -13,33 +13,32 @@
-            }
+			foreach (var eios in actorNode.ChildrenMatching("EmitInfantryOnSell"))
+			{
+				eios.RenameKeyPreservingSuffix("SpawnActorsOnSell");
```
Can we please change `RenameKeyPreservingSuffix` to `RenameKey(..., bool preserveSuffix=true)` for consistency?

pchote commented on OpenRA.Mods.Common/UpdateRules/UpdateUtils.cs:
```diff
@@ -272,6 +272,22 @@ public static void RemoveNode(this MiniYamlNode node, MiniYamlNode toRemove)
+		}
+
 		/// <summary>Removes children with keys equal to [match] or [match]@[arbitrary suffix]</summary>
 		public static int RemoveNodes(this MiniYamlNode node, string match)
```
We probably also want a `bool matchSuffix=true` parameter on these too for consistency.  Thankfully none of our the rules will need to change to go with this.

reaperrr committed:
Add some more useful extensions to UpdateUtils

In many situations, these will save up to 3-4 lines, because quite often this will be enough to avoid curly brackets, "if"s or one-by-on renamings.

reaperrr commented:
Updated.

Reviewing commit by commit recommended, gh messed up the order a little, 1db2565 is actually the last commit.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/df83f17508f4ebc674a33fbf56387c5d84a5626d...bfa4d5a94972bcee08c43263509ca562f2aea115)
